web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx

```diff
@@ -0,0 +1,155 @@
+  state: AgentProcessState;
+  runAgentAndWaitForNodeToJoin(): Promise<void>;
+}
+
+const ConnectMyComputerContext = createContext<ConnectMyComputerContext>(null);
+
+export const ConnectMyComputerContextProvider: FC<{
+  rootClusterUri: RootClusterUri;
+}> = props => {
+  const { mainProcessClient, connectMyComputerService } = useAppContext();
+  const [agentState, setAgentState] = useState<AgentProcessState>(
+    () =>
+      mainProcessClient.getAgentState({
+        rootClusterUri: props.rootClusterUri,
+      }) || {
+        status: 'not-started',
+      }
+  );
+
+  const runAgentAndWaitForNodeToJoin = useCallback(async () => {
+    await connectMyComputerService.runAgent(props.rootClusterUri);
+
+    // TODO(gzdunek): Replace with waiting for the node to join.
+    const waitForNodeToJoin = wait(1_000);
+
+    await Promise.race([
+      waitForNodeToJoin,
+      waitForAgentProcessErrors(
+        mainProcessClient.subscribeToAgentUpdate,
+        props.rootClusterUri
+      ),
+    ]);
+  }, [
+    connectMyComputerService,
+    mainProcessClient.subscribeToAgentUpdate,
+    props.rootClusterUri,
+  ]);
+
+  useEffect(() => {
+    const { cleanup } = mainProcessClient.subscribeToAgentUpdate(
+      props.rootClusterUri,
+      state => setAgentState(state)
+    );
+    return cleanup;
+  }, [mainProcessClient, props.rootClusterUri]);
+
+  return (
+    <ConnectMyComputerContext.Provider
+      value={{
+        state: agentState,
+        runAgentAndWaitForNodeToJoin,
+      }}
+      children={props.children}
+    />
+  );
+};
+
+export const useConnectMyComputerContext = () => {
+  const context = useContext(ConnectMyComputerContext);
+
+  if (!context) {
+    throw new Error(
+      'ConnectMyComputerContext requires ConnectMyComputerContextProvider context.'
+    );
+  }
+
+  return context;
+};
+
+/**
+ * Waits for `error` and `exit` events from the agent process.
+ * If none of them happen within 20 seconds, the promise resolves.
+ */
+async function waitForAgentProcessErrors(
+  subscribeToAgentUpdate: SubscribeToAgentUpdate,
+  rootClusterUri: RootClusterUri
+) {
+  let cleanupFn: () => void;
+
+  try {
+    const errorPromise = new Promise((_, reject) => {
+      const { cleanup } = subscribeToAgentUpdate(rootClusterUri, agentState => {
+        if (agentState.status === 'exited') {
+          const { code, signal } = agentState;
+          const codeOrSignal = [
+            // code can be 0, so we cannot just check it the same way as the signal.
+            code != null && `code ${code}`,
+            signal && `signal ${signal}`,
+          ]
+            .filter(Boolean)
+            .join(' ');
+
+          reject(
+            new Error(
+              [
+                `Agent process exited with ${codeOrSignal}.`,
+                agentState.stackTrace,
+              ]
+                .filter(Boolean)
+                .join('\n')
+            )
+          );
+        }
+        if (agentState.status === 'error') {
```

Comment:
The app doesn't seem to respond correctly to the error status.

I modified agentRunner like this:

```typescript
    let { agentBinaryPath } = this.settings;
    agentBinaryPath += '-elo';
```

I see that agent runner has updated the agent state (since I added a log statement), but the app doesn't report any errors and shows a success state.

If this is a bug on the side of agentRunner, let's make sure to cover it by a test case.